src/Workspaces/Core/Portable/Shared/Extensions/ISymbolExtensions.cs

```diff
@@ -488,58 +488,31 @@ public static bool IsUnsafe(this ISymbol member)
                 return type;
             }
 
-            if (symbol is IMethodSymbol method && !method.Parameters.Any(p => p.RefKind != RefKind.None))
+            if (symbol is IMethodSymbol method && method.Parameters.All(p => p.RefKind == RefKind.None))
             {
+                var count = extensionUsedAsInstance ? Math.Max(0, method.Parameters.Length - 1) : method.Parameters.Length;
+                var skip = extensionUsedAsInstance ? 1 : 0;
+
+                string WithArity(string typeName, int arity) => arity > 0 ? typeName + '`' + arity : typeName;
+
                 // Convert the symbol to Func<...> or Action<...>
-                if (method.ReturnsVoid)
-                {
-                    var count = extensionUsedAsInstance ? method.Parameters.Length - 1 : method.Parameters.Length;
-                    var skip = extensionUsedAsInstance ? 1 : 0;
-                    count = Math.Max(0, count);
-                    if (count == 0)
-                    {
-                        // Action
-                        return compilation.ActionType();
-                    }
-                    else
-                    {
-                        // Action<TArg1, ..., TArgN>
-                        var actionName = "System.Action`" + count;
-                        var actionType = compilation.GetTypeByMetadataName(actionName);
+                var delegateType = compilation.GetTypeByMetadataName(method.ReturnsVoid
+                    ? WithArity("System.Action", count)
+                    : WithArity("System.Func", count + 1));
 
-                        if (actionType != null)
-                        {
-                            var types = method.Parameters
-                                .Skip(skip)
-                                .Select(p =>
-                                    p.Type == null ?
-                                    compilation.GetSpecialType(SpecialType.System_Object) :
-                                    p.Type)
-                                .ToArray();
-                            return actionType.Construct(types);
-                        }
-                    }
-                }
-                else
+                if (delegateType != null)
                 {
-                    // Func<TArg1,...,TArgN,TReturn>
-                    //
-                    // +1 for the return type.
-                    var count = extensionUsedAsInstance ? method.Parameters.Length - 1 : method.Parameters.Length;
-                    var skip = extensionUsedAsInstance ? 1 : 0;
-                    var functionName = "System.Func`" + (count + 1);
-                    var functionType = compilation.GetTypeByMetadataName(functionName);
-
-                    if (functionType != null)
+                    var types = method.Parameters
+                        .Skip(skip)
+                        .Select(p => p.Type ?? compilation.GetSpecialType(SpecialType.System_Object));
```

Comment:
can this actually ever be null? is it possible this check was redundant?

the tests pass even if i remove this

Reply:
it's in the previous code too, but I'm not sure if it's actually useful...
on the other hand I don't want to fix one source of NullReferenceException and introduce another one :D